cmds/exp/nvme_unlock/nvme_unlock.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2022 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// The nvme_unlock command is used to unlock a NVMe drive with a
+// HSS-derived password, and rescan the drive to enumerate the
+// unlocked partitions.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"regexp"
+	"syscall"
+	"unsafe"
+
+	"github.com/u-root/u-root/pkg/hsskey"
+	"github.com/u-root/u-root/pkg/mount/block"
+)
+
+const (
+	opalLockUnlockIoctl = 1092120797
+)
+
+type opalKey struct {
+	lr     byte
+	keyLen byte
+	align  [6]byte
+	key    [256]byte
+}
+
+type opalSessionInfo struct {
+	sum     uint32
+	who     uint32
+	opalKey opalKey
+}
+
+type opalLockUnlock struct {
+	session opalSessionInfo
+	lState  uint32
+	align   [4]byte
+}
+
+var (
+	disk               = flag.String("disk", "/dev/nvme0n1", "The disk to be unlocked")
+	verbose            = flag.Bool("d", false, "print debug output")
+	verboseNoSanitize  = flag.Bool("dangerously-disable-sanitize", false, "Print sensitive information - this should only be used for testing!")
+	noRereadPartitions = flag.Bool("no-reread-partitions", false, "Only attempt to unlock the disk, don't re-read the partition table.")
+	lock               = flag.Bool("lock", false, "Lock instead of unlocking")
+)
+
+func verboseLog(msg string) {
+	if *verbose {
+		log.Print(msg)
+	}
+}
+
+func getSysfsInfo(index string, field string) (string, error) {
+	data, err := os.ReadFile(fmt.Sprintf("/sys/class/nvme/nvme%s/%s", index, field))
+	if err != nil {
+		return "", err
+	}
+	for i, b := range data {
+		if b == ' ' || b == '\n' {
+			return string(data[0:i]), nil
+		}
+	}
+	return string(data), nil
+}
+
+func run() error {
+	commandName := "unlock"
+	if *lock {
```

Comment:
I think the goal of creating a `run()` function is to pass in all the dependencies and flag values to make it easier for testing.

In this example, we'd probably want to add `lock` as a parameter, instead of referencing `*lock` here.

Basically any flag value (*value) should be accessed in `main()` only and passed to `run()` so that we can make it testable in the future.

Reply:
Fixed.